internal/driver/internal/k8s/logicalvolume_service.go

```diff
@@ -238,23 +241,23 @@ func (s *LogicalVolumeService) DeleteVolume(ctx context.Context, volumeID string
 	}
 
 	// wait until delete the target volume
-	for {
-		logger.Info("waiting for delete LogicalVolume", "name", lv.Name)
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-time.After(100 * time.Millisecond):
-		}
-
-		err := s.getter.Get(ctx, client.ObjectKey{Name: lv.Name}, new(topolvmv1.LogicalVolume))
-		if err != nil {
-			if apierrors.IsNotFound(err) {
-				return nil
+	return wait.ExponentialBackoffWithContext(ctx,
+		wait.Backoff{
+			Duration: 100 * time.Millisecond, // initial backoff
+			Factor:   2,                      // factor for duration increase
+			Jitter:   0.1,
+			Steps:    math.MaxInt, // run for infinity; we assume context gets canceled
+		}, func(ctx context.Context) (bool, error) {
```

Comment:
I removed caps and maximum backoff before abort after consideration with @toshipp. IMO too long backoff time cannot be avoided with wait.Backoff because if we reach the Cap it will get cancelled. I can only write our own backoff logic for this case or we accept a maximum timeout.

Reply:
Oh, I didn't know the behavior. I also think no-cap is not good, but I couldn't find other useful functions.

Reply:
How about this code? I think this effectively do an exponential backoff with the limited backoff time and unlimited retries.

```go
wait.Backoff{
	Duration: 100 * time.Millisecond,
	Factor:   2,
	Jitter:   0.1,
	Steps:    math.MaxInt,
	Cap:      5 * time.Minute,
}.DelayFunc().Until(ctx, false, false, func(ctx context.Context) (bool, error) {
	// check conditions
})
```

Reply:
This will stop retrying once 5 minutes are hit as duration. See https://github.com/kubernetes/apimachinery/blob/c9c3e94f52f06444120a09bfbdb11bec38a5c8e1/pkg/util/wait/backoff.go#L111 for the concrete logic that sets steps to 0

Reply:
Seriously? If I understand correctly, it is just a logic of ExponentialBackoffWithContext to stop retrying as Backoff.Steps becomes zero. Generally it's not the case.

Test Code:

```go
func TestExponentialBackoff(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var prev time.Time
	f := func(ctx context.Context) (bool, error) {
		if !prev.IsZero() {
			fmt.Printf("%s\n", time.Since(prev))
		}
		prev = time.Now()
		return false, nil
	}

	wait.Backoff{
		Duration: 100 * time.Millisecond,
		Factor:   2,
		Jitter:   0.1,
		Steps:    math.MaxInt,
		Cap:      1 * time.Second,
	}.DelayFunc().Until(ctx, true, false, f)
}
```

Output:

```
=== RUN   TestExponentialBackoff
400ns
110.298409ms
201.412217ms
424.604589ms
872.976143ms
1.019066153s
1.014166001s
1.077151559s
1.074724855s
1.039145566s
1.038260597s
1.038210446s
1.033878205s
--- PASS: TestExponentialBackoff (10.00s)
```

NOTE: The first retry after 400ns seems to be a bug that is already fixed in the master branch of k8s ([commit](https://github.com/kubernetes/kubernetes/commit/79f9bfdc8e5f5805c98741fd10e57db95c3c604e)).

Reply:
You are absolutely correct, I completely missed that you switched from ExponentialBackoffWithContext to DelayFunc. I will switch accordingly. Thanks for the correction

Reply:
Except for the spec update i now switched everything to the capped delay with a maximum of 10s. The spec update with conflict checking I left with the default backoff because we do not expect long wait times on this and should error if we have update contention.